designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,313 @@
+
+```js
+// Inside the rule tester class.
+function runRuleForItem(item) {
+  // ...
+
+  // Only surround these calls with a try/catch if the current test case is for a fatal error.
+  try {
+      validate(...);
+      // or
+      messages = linter.verify(...);
+  } catch (err) {
+      if (err.messageForTest) {
+          // Return a message so this exception can be tested.
+          return {
+              messages: [{
+                  ruleId: ruleName,
+                  fatal: true,
+                  message: err.messageForTest,
+                  name: err.name === 'Error' ? err.constructor.name : err.name,
+              }],
+          };
+      }
+
+      // Not one of the relevant exceptions for testing.
+      throw err;
+  }
+}
+```
+
+> Draft/WIP PR demonstrating a working implementation of the current design: [#16823](https://github.com/eslint/eslint/pull/16823)
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+
+We will document this change in the ESLint [RuleTester](https://eslint.org/docs/latest/developer-guide/nodejs-api#ruletester) section of the [Node.JS API](https://eslint.org/docs/latest/developer-guide/nodejs-api) page.
+
+Ideally, we can also draw attention to it with a paragraph and example usage in the blog post / release notes for the version it ships in. This is the least we can do to raise awareness of it for rule authors who may be interested to take advantage of it in new or existing rules.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+- A user testing that their rule schema catches invalid options may end up including the message from JSON Schema in their test case (e.g. `Value "bar" should be boolean.`). This could make it more difficult for ESLint to upgrade its version of ajv / JSON Schema in the future, as an upgraded version could tweak messages and necessitate updating any test cases that include the message. This is a relatively minor concern, as it's unlikely that messages will change often, and it's easy to update test cases if they do. Two factors that could mitigate this:
+  - If [snapshot testing](https://github.com/eslint/eslint/issues/14936) is implemented in the future, it would become possible to fix test cases with a single command.
+  - We can encourage the use of `error.name` property (discussed in [Detailed Design](#detailed-design)) for the majority of cases that don't care about including the exact exception text.
```

Comment:
I'm happy to document this. Note that the error message changing should be rare, only if:
- The user changes the error message in their own rule
- An upgraded version of ajv changes the validation text

Reply:
> * The user changes an error message in their own rule

I was thinking to add this note only for schema validation errors as those are the messages we produce.

> * An upgraded version of ajv changes the validation text

That should be really rare, but still we can't guarantee it won't happen.

Reply:
Seems we agree on this. Can you add a note to the RFC document so that we don't forget this? Perhaps in the [Documentation](https://github.com/eslint/rfcs/pull/103/files#diff-f782a5c8d8f504e2963c007c44dedd48c24cd73e6d7fb939e8d7ebe80b59de91R177) section.

Reply:
I added a note about documenting this in the "Drawbacks" section where I describe this issue.